### PR TITLE
PixelPaint: Polygonal select tool

### PR DESCRIPTION
**Pull request by tslater2006**
This adds a polygonal selection tool where the user can make an initial click for the starting point of the shape and then subsequent clicks add vertexes. Double clicking completes the shape by connecting the last point to the first one. Pressing escape during selection cancels the selection. Details of how the mask is generated are below. 

This PR also includes a few smaller commits that address some bugs in the Wand Tool and the "Delete Selection" logic. For the Wand Tool, there was an issue of the selection not accounting for the layer being offset from (0,0).

Wand Tool also wasn't listening for the Esc key to clear the selection. 

For the "Delete Selection" bug, a selection can go outside of the layer bounds (rectangle selection, and polygon selection), attempting to delete all pixels in the selection was not checking to see if the selection was in the bitmap for the layer.

The polygonal mask is generated as follows:

- First we calculate the size of the bounding rect needed to hold the
  polygon
- We add 2 pixels to height/width to allow us a 1 pixel border, the
  polygon will be centered in this bitmap
- Draw the polygon into the bitmap via Gfx::Painter, making sure to
  connect final polygon point to the first to ensure an enclosed shape
- Generate a selection mask the size of the bitmap, with all pixels
  initially selected
- Perform a flood fill from (0,0) which is guaranteed to be outside the
  polygon
- For every pixel reached by the flood fill, we clear the selected pixel
  from the selection mask
- Finally we merge the selection mask like other selection tools.

### CODE REVIEW

**Comment by tslater2006** [comment]
I've rebased this on latest master.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:58]: This flood fill is very similar to the ones used in your other tools, could it be factored out into a shared function?

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:101]: Tiny thing: I think you might be able to just translate the painter and then paint your polygon directly, which might be slightly simpler.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:131]: Super pedantic nit: This comment and a few don't match the serenity style (proper sentences,  capitalization, full-stops)

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:184]: Nit: You should have the braces around this else branch as you have them in the if/then branch.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:192]: Tiny nit: You can remove the braces here

**Review by MacDue — COMMENTED** [review]
Few little things. Think this looks very nice :cowboy_hat_face:

**Comment by MacDue** [comment]
If you've not post this in `#code-review` again to try to get some maintainers eyes on it. It's easy for PRs to get lost after a while.

**Comment by tslater2006** [comment]
Thanks for taking the time to go through this. I posted in code-review yesterday. I will work on addressing these tonight/tomorrow.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:58]: The honest answer here is that I don't know. It feels like it should though. Just off top of my head, the key difference here is what happens when you reach a pixel. In the bucket tool you color the pixel, in the Polygon Select Tool you toggle a point in a mask.

Maybe some abstracted "flood fill" class/method where you pass it a lambda of what to do for each pixel it reaches? I'm not sure where a good place to put such a thing would be. What about adding flood_fill to Bitmap itself in LibGFX and accepting a lambda for the action to perform on each pixel reached?

Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:58]: @MacDue I have an implementation of this locally and it seems to work just fine. It cleans up the tool code *a bunch*. I may extract this change into a separate branch and create a separate PR for it, and then rebase this (and another selection tool I'm working on) once that lands, or some variation of it :).

The abstraction approach can be discussed in that PR but to give an idea of what I currently have working, here's a couple snippets.

I currently have a function in Bitmap
```
void flood_visit_from_point(Gfx::IntPoint const& start_point, int threshold,Optional<Color> target_color, Function<void(Gfx::IntPoint location)> pixel_reached);
```
And then in a tool like Bucket Tool:

```
auto location_reached = [&](Gfx::IntPoint location) {
    bitmap.set_pixel<Gfx::StorageFormat::BGRA8888>(location.x(), location.y(), fill_color);
};

bitmap.flood_visit_from_point(start_position, threshold,target_color ,move(location_reached));
```

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:101]: Fixed!

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: Fixed!

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: Fixed!

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: Fixed!

**Comment by linusg** [comment]
@tslater2006 this has conflicts unfortunately, please rebase :)

**Comment by tslater2006** [comment]
I'm out of town for the next few days but as soon as I get back I will rebate and update this to use the just-landed flood fill in LibGfx. Thanks!

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:3]: I don't think this file was worked on by Kling in 2021 :^)

You are free (and encouraged) to add your own copyright line

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h:3]: Same here :^)

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:223]: Note: You can make a lookup like this entirely at compile time like so. Which avoids allocating a vector and some strings.

```
static constexpr auto s_merge_mode_names = []{
    Array<StringView, (int)Selection::MergeMode::__Count> names;
    for (size_t i = 0; i < names.size(); i++) {
        switch((Selection::MergeMode)i) {
            case Selection::MergeMode::Set:
                names[i] = "Set";
                break;
             // ...
        }
    }
    return names;
}();
```

Where this is just placed somewhere in the .cpp.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:101]: You said you fixed my previous comment here but I see no change, perhaps you forgot to push it?

**Review by MacDue — APPROVED** [review]
Few minor things, but I'd like to see this merged :^)

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:3]: Copy paste strikes again. I'll get this fixed on Monday

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:101]: Hmmm, I'll take a look. Sorry about that, I'll get it fixed properly

Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: I just force pushed after rebasing and resolving some other issues. This should be fixed now.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h]: Fixed :)

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: Fixed :)

**Comment by tslater2006** [comment]
Rebased and conflicts resolved @linusg.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (MacDue) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:198]: That's not true. `GUI::ItemListModel` is a template that can accept any container type, which by default is a `Vector`, and you've currently specialized it for `AK::String`. To initialize one using the static array of `StringView`s you'd do something like:

```c++
    auto model = GUI::ItemListModel<StringView, decltype(s_merge_mode_names)>::create(s_merge_mode_names);
```

(I'm using `decltype` since that's more convenient than writing the type)

See full example:  https://godbolt.org/z/Pb7YsnjrG

Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:198]: Hey thanks! Can you tell I'm very very new to C++? :) I'll get this implemented this evening and push the changes.

Comment (MacDue) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:198]: It's fine, this was more of an optional suggestion anyway :)

Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: This has been implemented, thanks again for the help getting it working.

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp]: Marking this resolved since the extraction of flood fill logic has landed in master and most recent push of this PR leverages the new code :)

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (linusg) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:122]: Please add a newline between functions

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (linusg) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:130]: Likewise

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h
Comment (linusg) [inline Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h:11]: No need to separate these

**Review by linusg — COMMENTED** [review]
Looks good, just a few style nitpicks

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp:123]: Fixed!

File: Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.h]: Fixed!

**Pull request merged**
